### PR TITLE
fix(filtered-deck): restore state on recreation (#21463)

### LINKED ISSUE #21463: Filtered deck options: name validation and discard-changes dialog stop working after the screen is restored
### Checked for duplicates?

- [x] This issue is not a duplicate

### Does it also happen in the desktop version?

- [x] This bug does not occur in the latest version of Anki Desktop

### What are the steps to reproduce this bug?

1. enable "Don't keep activities" from developer options
2. open filtered deck options screen (either make a new filtered deck or open existing one)
3. go to some other app and come back, so the screen gets recreated
4. then try any of these two:
   a) type a name of some deck which already exist
   b) change anything like search or limit, then press back

and theres one more which dont even need process death:
5. open a existing filtered deck options, rename it to anything else, then type the original name
   back again

### Expected behaviour

for (a) it should show that "name already exists" error and Build button should stay disabled, for
(b) it should ask me discard changes before deleting my edits, and in (5) putting back the original
name shouldnt be counted as conflict since its the same deck.

but none of these works. after the screen is restored, typing a existing deck name shows no error
at all and Build stays enabled, so you end up getting some raw backend exception. and back press
just throws away everything without asking anything. and reverting the name marks it as "already
exists" and Rebuild gets disabled.

i wrote few unit tests before reporting to make sure im not imagining it, all 3 fails on main:

  DEFECT 1 (restore + existing name)  -> expected: <AlreadyExists> but was: <null>
  DEFECT 2 (restore + edit)           -> Expected value to be true.
  DEFECT 3 (revert to own name)       -> expected: <null> but was: <AlreadyExists>

(for simulating the process death i just created a second ViewModel on the same SavedStateHandle,
cus thats what basically happens when VM is killed but handle survives)

i think the reason is this early return in FilteredDeckOptionsViewModel's init:

    val previousState = savedStateHandle.get<FilteredDeckOptions>(ARG_DATA)
    if (previousState != null) {
        state.update { previousState }
        return@launch
    }

because of this return it skips two lines below - decksNames = withCol { safeGetDecksNames() } and
initialState = this. so decksNames remains emptyList() and thats why decksNames.contains(name)
inside onDeckNameChange never becomes true. and initialState remains null so wasStateModified()
hits `val initial = initialState ?: return false` and returns false always, which means
hasUnsavedChanges is always false and the back handler never gets enabled.

for the 3rd one, safeGetDecksNames() is using includeFiltered = true so the list also contain the
same deck which you are editing right now, thats why setting original name back collides with
itself

### Debug info

```text
AnkiDroid Version = 2.24.0 (ebcf8e0e34921628b9b8a496c66ffd4adbb3705f)  
Backend Version = 0.1.64-anki25.09.2 (25.09.2 3890e12c9e48c028c3f12aa58cb64bd9f8895e30)  
Android Version = 13 (SDK 33)  
ProductFlavor = play  
Device Info = OnePlus | OnePlus | OP555BL1 | IV2201_IND | IV2201 | mt6877  
WebView Info = [com.google.android.webview | 787118103]: Mozilla/5.0 (Linux; Android 13; IV2201 Build/TP1A.220905.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/150.0.7871.181 Mobile Safari/537.36  
ACRA UUID = bbe8603c-4564-4f9f-9621-e2c7e5b9b556  
FSRS = 5.1.0 (Enabled: false)  
Crash Reports Enabled = true
```

### (Optional) Anything else you want to share?

_No response_

### Research

- [x] I have checked the [manual](https://docs.ankidroid.org/) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by GuTS805**
this fixes the 3 things from the issue - after the screen gets recreated the duplicate name check
and the discard changes dialog both stop working, plus renaming a filtered deck back to its own
name was getting flagged as duplicate.

* Fixes #21463

## Purpose / Description
the init block in FilteredDeckOptionsViewModel was returning early whenever there was a previous
state saved:

    val previousState = savedStateHandle.get<FilteredDeckOptions>(ARG_DATA)
    if (previousState != null) {
        state.update { previousState }
        return@launch
    }

problem is two things below that return never ran. decksNames stayed as emptyList() so the
decksNames.contains(name) check inside onDeckNameChange could never be true, and initialState
stayed null so wasStateModified() hit `val initial = initialState ?: return false` and always
returned false. so after a restore you could type a name that already exist and get no error at
all (Build stays enabled, you just get a raw backend exception when you press it), and pressing
back would throw away all your edits without asking anything.

the third one is unrelated to restore - safeGetDecksNames() uses includeFiltered = true so the
list also has the deck you are editing right now in it. so if you rename an existing filtered deck
and then put the original name back it says "already exists" and disables Rebuild.

## Approach
moved `decksNames = withCol { safeGetDecksNames() }` above the early return since both paths need
it. for initialState i couldnt just read it back from ARG_DATA because that key holds the edited
state, not the original one, so its saved under its own key (ARG_INITIAL_DATA) and read back on
the restore path.

for the third one i used the existing `title` field - its already documented as the name from when
the deck was first loaded and it doesnt change for the lifetime of the screen, so its exactly what
was needed. only applied it when id != null so it doesnt affect creating new decks.

## How Has This Been Tested?
added 3 tests to FilteredDeckOptionsViewModelTest, all of them failed before the fix:

  name validation still works after the screen is restored -> expected <AlreadyExists> but was <null>
  changed status still works after the screen is restored  -> Expected value to be true
  reverting to the deck's own name is not a duplicate      -> expected <null> but was <AlreadyExists>

for simulating the restore theres a small withRestoredViewModel helper - it builds a view model,
throws it away and builds a second one over the same SavedStateHandle, which is basically what
happens when the VM dies but the handle survives.

also did mutation testing to make sure none of the 3 changes are redundant:
- removed just the initialState restore -> only test 2 failed
- removed the decksNames move + the own name check -> only tests 1 and 3 failed

whole com.ichi2.anki.filtered package is green after (17 tests), ktlint passes.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt [hunk 1]
```diff
@@ -70,8 +70,11 @@ class FilteredDeckOptionsViewModel(
     init {
         viewModelScope.launch {
             Timber.i("Starting filtered deck options setup, deckId=$did")
+            // needed on both paths: name validation is done against this list
+            decksNames = withCol { safeGetDecksNames() }
             val previousState = savedStateHandle.get<FilteredDeckOptions>(ARG_DATA)
             if (previousState != null) {
+                initialState = savedStateHandle[ARG_INITIAL_DATA]
                 state.update { previousState }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt:78]: I believe this needs to update `hasUnsavedChanges`

If so, could you add a regression test

Comment (GuTS805) [reply AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt:78]: It was my bad. You were right. hasUnsavedChanges isnt in the SavedStateHandle so restoring
initialState alone wasnt enough edits made before the process died werent getting reported
after. my earlier test only covered edits after the restore, thats why i missed it.

added the recompute on the restore path + a regression test for it. confirmed it fails without
the line

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt:78]: No worries, it happens! Truly appreciate all the effort you've been putting in

**Review by david-allison — APPROVED** [review]

**Review by david-allison — APPROVED** [review]
Thanks!

No need to re-request a review unless you feel it's required. 

It partially removes the 'approved' in the GitHub UI.

**Comment by GuTS805** [comment]
Oh! sorry. I tapped it by mistake..

**Review by criticalAY — APPROVED** [review]
LGTM! thanks

**Comment by criticalAY** [comment]
- rebased, thanks!

**Comment by GuTS805** [comment]
Thanks for the review ...
